src/core/tensor/tensor_math_cuda.h

```diff
@@ -20,260 +20,428 @@
+#include "singa/core/tensor.h"
 #include "./tensor_math.h"
 #include "./math_kernel.h"
 #include "singa/utils/cuda_utils.h"
 #include "singa/core/common.h"
 #include <cuda_runtime.h>
 #include <cublas_v2.h>
 #include "singa/utils/cuda_utils.h"
+#include <cudnn.h>
 
 namespace singa {
 
 /// out[i] = |in[i]|
 template <>
-void Abs<float, lang::Cuda>(const size_t num, const Block* in, Block* out,
+void Abs<float, lang::Cuda>(const Tensor* in, Tensor* out,
                             Context* ctx) {
-  const float* inPtr = static_cast<const float*>(in->data());
-  float* outPtr = static_cast<float*>(out->mutable_data());
+  const float* inPtr = static_cast<const float*>(in->block()->data());
+  float* outPtr = static_cast<float*>(out->block()->mutable_data());
+  const size_t num = in->Size();
   cuda::abs(num, inPtr, outPtr, ctx->stream);
 }
+
+template <>
+void Set<float, lang::Cuda>(const float x, Tensor* out,
+                            Context* ctx) {
+  float* outPtr = static_cast<float*>(out->block()->mutable_data());
+  float valuePtr[1] = {x};
+  int dimOut[out->shape().size()+1]; int stridesOut[out->strides().size()+1];
+  out->shape_cuda(dimOut,8);
+  out->strides_cuda(stridesOut,8);
+
+  cudnnDataType_t cudnn_dtype = CUDNN_DATA_FLOAT;
+  cudnnTensorDescriptor_t out_desc;
+  cudnnCreateTensorDescriptor(&out_desc);
+  cudnnSetTensorNdDescriptor(out_desc, cudnn_dtype, out->nDim(), dimOut, stridesOut);
+  cudnnSetTensor(ctx->cudnn_handle, out_desc, outPtr, valuePtr);
+
+  cudnnDestroyTensorDescriptor(out_desc);
+}
+
 /// out = in + x
+// template <>
+// void Add<float, lang::Cuda>(const Tensor* in, const float x,
+//                             Tensor* out, Context* ctx) {
+//   const float* inPtr = static_cast<const float*>(in->block()->data());
+//   float* outPtr = static_cast<float*>(out->block()->mutable_data());
+//   cuda::add(num, inPtr, x, outPtr, ctx->stream);
+// }
+
 template <>
-void Add<float, lang::Cuda>(const size_t num, const Block* in, const float x,
-                            Block* out, Context* ctx) {
-  const float* inPtr = static_cast<const float*>(in->data());
-  float* outPtr = static_cast<float*>(out->mutable_data());
-  cuda::add(num, inPtr, x, outPtr, ctx->stream);
+void Add<float, lang::Cuda>(const Tensor* in, const float x,
+                            Tensor* out, Context* ctx) {
+  Set<float, lang::Cuda>(x, out, ctx);
+  const float* inPtr = static_cast<const float*>(in->block()->data());
+  float* outPtr = static_cast<float*>(out->block()->mutable_data());
+  int dimIn[in->shape().size()+1]; int stridesIn[in->strides().size()+1];
+  int dimOut[out->shape().size()+1]; int stridesOut[out->strides().size()+1];
+  in->shape_cuda(dimIn,8);
+  in->strides_cuda(stridesIn,8);
+  out->shape_cuda(dimOut,8);
+  out->strides_cuda(stridesOut,8);
+
+  float alpha = 1.0, beta=1.0;
+  cudnnDataType_t cudnn_dtype = CUDNN_DATA_FLOAT;
+  cudnnTensorDescriptor_t in_desc, out_desc;
+  cudnnCreateTensorDescriptor(&in_desc);
```

Comment:
since the code for setup the cudnn data structures for a Tensor is the same for tensors, why not create a function to do the setup, e..g

    create_cudnn_tensor(const Tensor& x){
      CUDNN_CHECK(cudnnCreateTensorDescirptor());
      CUDNN_CHECK(cudnnSetTensorNdDescriptor());
    }

Reply:
does it make sense?